### PR TITLE
SpecieInfo screen & Syncing User species and species' Images

### PR DESCRIPTION
**Pull request by tejassonar**
Changes in this pull request:
- SpecieInfo Screen 
- Syncing of species with backend 
- Saving Specie Image in the file system and syncing

To-Do
- handle crash due to https://github.com/Plant-for-the-Planet-org/treemapper/pull/270#issuecomment-811983129

### CODE REVIEW

File: app/components/ManageSpecies/SpecieInfo.js [hunk 1]
```diff
@@ -1,21 +1,160 @@
...
+  let SpecieDescription = route.params.SpecieDescription;
+
+  useEffect(() => {
+    setAliases(route.params.SpecieAliases);
+    if (route.params.SpecieLocalImage) {
+      setLocalImageUrl(route.params.SpecieLocalImage);
+    }
+    if (route.params.SpecieImage && !route.params.SpecieLocalImage)
+      RNFS.downloadFile({
+        fromUrl: `https://bucketeer-894cef84-0684-47b5-a5e7-917b8655836a.s3.eu-west-1.amazonaws.com/development/media/cache/species/default/${route.params.SpecieImage}`,
```
Comment (sagararyal) [inline app/components/ManageSpecies/SpecieInfo.js:40]: this should not be hardcoded.

Please use cdn url to fetch files like this

Comment (tejassonar) [reply app/components/ManageSpecies/SpecieInfo.js:40]: Yes, I forgot to add CDN. I will add it in the current commit

Comment (norbertschuler) [reply app/components/ManageSpecies/SpecieInfo.js:40]: compare https://github.com/Plant-for-the-Planet-org/treemapper/blob/86665a75f748e6bdc2cf62a6f67452b82ca03f11/app/components/MainScreen/index.js#L245

**Comment by norbertschuler** [comment]
- [ ] After I entered an alias name and pressed "Enter" in my iOS simulator, the app crashed with
```
2021-03-29 12:19:18.609744+0200 Tree Mapper[32523:1329221] [javascript] { aliases: 'Test Tree', description: undefined }, 'data'
libc++abi.dylib: terminating with uncaught exception of type realm::js::TypeErrorException
terminating with uncaught exception of type realm::js::TypeErrorException
CoreSimulator 732.18.6 - Device: iPhone 11 (6186C773-F412-45B4-8FB7-C22CE411C02C) - Runtime: iOS 14.4 (18D46) - DeviceType: iPhone 11
(lldb) 
```
Also see Bugsnag at https://app.bugsnag.com/plant-for-the-planet/tree-mapper/errors/6061a8acca91430007db0f0d?filters[event.since][0]=30d&filters[error.status][0]=new

Simulator Screen Shot - iPhone 11 - 2021-03-29 at 12 19 15

**Comment by norbertschuler** [comment]
- [ ] For now the picture can only be changed once, right? As I have found not function to upload another one or delete the existing one yet?

Simulator Screen Shot - iPhone 11 - 2021-03-29 at 12 27 08

- [ ] I got this error after taking a photo for the species:

[Mon Mar 29 2021 12:24:31.753]  LOG      file:///Users/norbertschuler/Library/Developer/CoreSimulator/Devices/6186C773-F412-45B4-8FB7-C22CE411C02C/data/Containers/Data/Application/89C3A3C2-CBE5-4ABC-86C0-FEC7BA1445E8/Library/Caches/Camera/44875EDE-BA3E-4F88-9E2F-23E0FEEB7BD5.jpg Image url
[Mon Mar 29 2021 12:24:31.778]  LOG      44875EDE-BA3E-4F88-9E2F-23E0FEEB7BD5.jpg fsurl
[Mon Mar 29 2021 12:24:33.403]  LOG      [Error: Request failed with status code 401] create error
[Mon Mar 29 2021 12:24:33.406]  LOG      [Error: Request failed with status code 401] UpdateSpeciesImage Error

**Comment by norbertschuler** [comment]
- [ ] Trying to change a species description I get an OAuth error (I logged in a few days before). But the token does not automatically refreshes itself and I get this OAuth0 error again and again here:

```
[Mon Mar 29 2021 12:30:20.336]  LOG      {"aliases": "", "description": "Is this visible to everybody?"} data
[Mon Mar 29 2021 12:30:22.317]  ERROR    Error at /actions/species/setAliases, {"data":"OAuth Token is invalid or expired","status":401,"headers":{"content-type":"application/json","alt-svc":"h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400","cache-control":"no-cache, private","cf-request-id":"091f21e3c4000041624712b000000001","via":"1.1 vegur","cf-cache-status":"DYNAMIC","expect-ct":"max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"","cf-ray":"63786c19381f4162-HAM","date":"Mon, 29 Mar 2021 10:30:22 GMT","strict-transport-security":"max-age=2592000; includeSubDomains; preload","server":"cloudflare","vary":"Accept-Encoding"},"config":{"url":"/treemapper/species/null","method":"put","data":"{\"aliases\":\"\",\"description\":\"Is this visible to everybody?\"}","headers":{"Accept":"application/json, text/plain, */*","Content-Type":"application/json","Authorization":"OAuth OKtMWOFnXY9yh9uyRxEIVF6Jy3Z8jX3q","X-SESSION-ID":"1f86c46d-f946-4423-be52-ba676bc47ea3"},"baseURL":"https://app-development.plant-for-the-planet.org","transformRequest":[null],"transformResponse":[null],"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1},"request":{"UNSENT":0,"OPENED":1,"HEADERS_RECEIVED":2,"LOADING":3,"DONE":4,"readyState":4,"status":401,"timeout":0,"withCredentials":true,"upload":{},"_aborted":false,"_hasError":false,"_method":"PUT","_perfKey":"network_XMLHttpRequest_https://app-development.plant-for-the-planet.org/treemapper/species/null","_response":"\"OAuth Token is invalid or expired\"","_url":"https://app-development.plant-for-the-planet.org/treemapper/species/null","_timedOut":false,"_trackingName":"unknown","_incrementalEvents":false,"responseHeaders":{"Content-Type":"application/json","Alt-Svc":"h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400","Cache-Control":"no-cache, private","cf-request-id":"091f21e3c4000041624712b000000001","Via":"1.1 vegur","cf-cache-status":"DYNAMIC","expect-ct":"max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"","cf-ray":"63786c19381f4162-HAM","Date":"Mon, 29 Mar 2021 10:30:22 GMT","Strict-Transport-Security":"max-age=2592000; includeSubDomains; preload","Server":"cloudflare","Vary":"Accept-Encoding"},"_requestId":null,"_headers":{"accept":"application/json, text/plain, */*","content-type":"application/json","authorization":"OAuth OKtMWOFnXY9yh9uyRxEIVF6Jy3Z8jX3q","x-session-id":"1f86c46d-f946-4423-be52-ba676bc47ea3"},"_responseType":"","_sent":true,"_lowerCaseResponseHeaders":{"content-type":"application/json","alt-svc":"h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400","cache-control":"no-cache, private","cf-request-id":"091f21e3c4000041624712b000000001","via":"1.1 vegur","cf-cache-status":"DYNAMIC","expect-ct":"max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"","cf-ray":"63786c19381f4162-HAM","date":"Mon, 29 Mar 2021 10:30:22 GMT","strict-transport-security":"max-age=2592000; includeSubDomains; preload","server":"cloudflare","vary":"Accept-Encoding"},"_subscriptions":[],"responseURL":"https://app-development.plant-for-the-planet.org/treemapper/species/null"}}
```

**Comment by norbertschuler** [comment]
> * [ ]  Trying to change a species description I get an OAuth error (I logged in a few days before). But the token does not automatically refreshes itself and I get this OAuth0 error again and again here:

- [ ] After I logged out and in again to refresh my OAuth0 token I did the same request and got a 500 error from the backend:

```
[Mon Mar 29 2021 12:34:55.762]  LOG      {"aliases": "", "description": "Is this visible to everybody else?"} data
[Mon Mar 29 2021 12:34:56.871]  ERROR    Error at /actions/species/setAliases, {"data":"Expected argument of type \"string\", \"null\" given at property path \"aliases\".\nfile: /app/vendor/symfony/property-access/PropertyAccessor.php, line: 198","status":500,"headers":{"content-type":"application/json","alt-svc":"h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400","cache-control":"max-age=0, must-revalidate, private","expires":"Mon, 29 Mar 2021 10:34:56 GMT","cf-request-id":"091f2617c000004175f81dc000000001","via":"1.1 vegur","cf-cache-status":"DYNAMIC","date":"Mon, 29 Mar 2021 10:34:56 GMT","cf-ray":"637872d2ce364175-HAM","expect-ct":"max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"","strict-transport-security":"max-age=2592000; includeSubDomains; preload","server":"cloudflare","vary":"Accept-Encoding"},"config":{"url":"/treemapper/species/spec_yI86RxVXJ68l","method":"put","data":"{\"aliases\":\"\",\"description\":\"Is this visible to everybody else?\"}","headers":{"Accept":"application/json, text/plain, */*","Content-Type":"application/json","Authorization":"OAuth ***","X-SESSION-ID":"1f86c46d-f946-4423-be52-ba676bc47ea3"},"baseURL":"https://app-development.plant-for-the-planet.org","transformRequest":[null],"transformResponse":[null],"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"maxBodyLength":-1},"request":{"UNSENT":0,"OPENED":1,"HEADERS_RECEIVED":2,"LOADING":3,"DONE":4,"readyState":4,"status":500,"timeout":0,"withCredentials":true,"upload":{},"_aborted":false,"_hasError":false,"_method":"PUT","_perfKey":"network_XMLHttpRequest_https://app-development.plant-for-the-planet.org/treemapper/species/spec_yI86RxVXJ68l","_response":"\"Expected argument of type \\\"string\\\", \\\"null\\\" given at property path \\\"aliases\\\".\\nfile: /app/vendor/symfony/property-access/PropertyAccessor.php, line: 198\"","_url":"https://app-development.plant-for-the-planet.org/treemapper/species/spec_yI86RxVXJ68l","_timedOut":false,"_trackingName":"unknown","_incrementalEvents":false,"responseHeaders":{"Content-Type":"application/json","Alt-Svc":"h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400","Cache-Control":"max-age=0, must-revalidate, private","Expires":"Mon, 29 Mar 2021 10:34:56 GMT","cf-request-id":"091f2617c000004175f81dc000000001","Via":"1.1 vegur","cf-cache-status":"DYNAMIC","Date":"Mon, 29 Mar 2021 10:34:56 GMT","cf-ray":"637872d2ce364175-HAM","expect-ct":"max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"","Strict-Transport-Security":"max-age=2592000; includeSubDomains; preload","Server":"cloudflare","Vary":"Accept-Encoding"},"_requestId":null,"_headers":{"accept":"application/json, text/plain, */*","content-type":"application/json","authorization":"OAuth ***","x-session-id":"1f86c46d-f946-4423-be52-ba676bc47ea3"},"_responseType":"","_sent":true,"_lowerCaseResponseHeaders":{"content-type":"application/json","alt-svc":"h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400","cache-control":"max-age=0, must-revalidate, private","expires":"Mon, 29 Mar 2021 10:34:56 GMT","cf-request-id":"091f2617c000004175f81dc000000001","via":"1.1 vegur","cf-cache-status":"DYNAMIC","date":"Mon, 29 Mar 2021 10:34:56 GMT","cf-ray":"637872d2ce364175-HAM","expect-ct":"max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"","strict-transport-security":"max-age=2592000; includeSubDomains; preload","server":"cloudflare","vary":"Accept-Encoding"},"_subscriptions":[],"responseURL":"https://app-development.plant-for-the-planet.org/treemapper/species/spec_yI86RxVXJ68l"}}
```

**Comment by norbertschuler** [comment]
- [ ] I just changed from reviewing #290 to this PP, reinstalled the app and logged in. After the login, the app got stuck repeating these lines endlessly in the console:

```
[Thu Apr 01 2021 16:51:06.596]  LOG      {"aliases": "", "description": ""} data
[Thu Apr 01 2021 16:51:06.598]  LOG      {"aliases": "", "description": ""} data
[Thu Apr 01 2021 16:51:06.602]  LOG      {"aliases": "", "description": ""} data
[Thu Apr 01 2021 16:51:06.603]  LOG      {"aliases": "", "description": ""} data
...
```

**Comment by norbertschuler** [comment]
> * [x]  For now the picture can only be changed once, right? As I have found not function to upload another one or delete the existing one yet?

- [ ] Now doing the photo again by clicking on the old one, I get a 500 error from the backend API call PUT https://app-development.plant-for-the-planet.org/treemapper/species/spec_yI86RxVXJ68l with `file: /app/src/MessageHandler/CommandBus/Species/SpeciesPatchHandler.php, line: 52`

**Comment by norbertschuler** [comment]
We should check the species sync against the use cases given in #291.

**Comment by norbertschuler** [comment]
- [ ] While testing I went back from choosing species to the screen taking a picture and the clicked "continue" again and got this crash ("Image can not be copied as an object with the same name already exists."):

```
AppContainer@http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false:67652:22
[Tue Apr 06 2021 11:15:58.918]  ERROR    error while saving file [Error: „C2D7F2E7-7E34-4EDF-9B8B-3D9BC3887452.jpg“ konnte nicht nach „Documents“ kopiert werden, da bereits ein Objekt mit demselben Namen existiert.]
[Tue Apr 06 2021 11:15:59.900]  ERROR    error while saving file [TypeError: undefined is not an object (evaluating '_ref18.inventory_id')]
```

**Comment by norbertschuler** [comment]
- [ ] I got this error trying to change a species on my real iPhone X device:

```
[Wed Apr 07 2021 12:53:01.296]  WARN     Non-serializable values were found in the navigation state. Check:

Main > SpecieInfo > params.toggleUserSpecies (Function)

This can break usage such as persisting and restoring state. This might happen if you passed non-serializable values such as function, class instances etc. in params. If you need to use components with callbacks in your options, you can use 'navigation.setOptions' instead. See https://reactnavigation.org/docs/troubleshooting#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state for more details. 

```

File: app/components/ManageSpecies/SpecieInfo.js [hunk 2]
```diff
@@ -1,21 +1,205 @@
...
+                </View>
+              )}
+              {/* <InfoCard /> */}
+              <View style={{ flex: 1, flexDirection: 'column' }}>
+                <Text style={styles.infoCardHeading}>Specie Name</Text>
+                <Text style={styles.infoCardText}>{specieName}</Text>
+                <Text style={styles.infoCardHeading}>Description</Text>
+                <TextInput
+                  style={[styles.infoCardText, { padding: 0 }]}
+                  placeholder={'Type description here'}
```
Comment (norbertschuler) [inline app/components/ManageSpecies/SpecieInfo.js:188]: Please add translations for these three text resources.

**Comment by norbertschuler** [comment]
- [ ] Some more padding to the left and right for the buttons and the back arrow (as on other screens) would be nice:

Simulator Screen Shot - iPhone 11 - 2021-04-08 at 17 11 58Simulator Screen Shot - iPhone 11 - 2021-04-08 at 17 12 26

**Comment by norbertschuler** [comment]
I successfully tested the sync and cleanup of species data if a user want to log out. 
- [ ] But what about the case of an accidentally logout due to an expired OAuth token? What would happen then with the locally remaining species data if you login to another account?

Bildschirmfoto 2021-04-08 um 16 55 59

**Comment by norbertschuler** [comment]
> * [ ]  But what about the case of an accidentally logout due to an expired OAuth token? What would happen then with the locally remaining species data if you login to your own or another account?

I assume all local species data is cleared as an 401 error triggers `auth0Logout` which again calls `resetAllSpecies`.

**Review by norbertschuler — APPROVED** [review]
LGFM

**Pull request merged**